// device-db/device-db.js
'use strict';

const moment = require('moment');
const Thing = require('thing');

const DeviceDb = function() {
  let deviceDb = Object.create(DeviceDb.prototype);
  deviceDb.devices = {};
  deviceDb.activityStream = [];
  return deviceDb;
};

DeviceDb.prototype.setDevice = function(deviceData) {
  let deviceId = deviceData.deviceid;
  let device;

  console.log('SET DEVICE', deviceData);

  if (!this.devices[deviceId]) {
    device = Thing(deviceData);
    this.devices[deviceId] = device;
  }
};

DeviceDb.prototype.getDevice = function(id) {
  return this.devices[id];
};

DeviceDb.prototype.getDeviceAttributes = function(id) {
  return this.devices[id].attributes;
};

DeviceDb.prototype.updateDeviceAttribute = function(update) {
  let device = this.devices[update.device];

  let updateText = device.update(update);
  console.log('UPDATETEXT', updateText);

  if (updateText !== false) {
    this.activityStream.unshift({
      id: update.device,
      attribute: update.attribute,
      value: update.value,
      timestamp: moment.utc(update.date.time),
      name: device.name,
      text: updateText
    });
  }

  console.log(this.activityStream);
};

module.exports = DeviceDb;
